templates/middleware-usage-reportquery.yaml

```
@@ -25,19 +32,108 @@ spec:
+    ),
+    cte_hourly_pod_by_product_label AS (
+      SELECT
+          L.labels['pod'] as pod,
+          L.labels['namespace'] as namespace,
+          L.labels['label_com_redhat_product'] as product,
+          date_trunc('hour', L."timestamp") as interval_start
+      FROM {| dataSourceTableName .Report.Inputs.KubePodLabelsComRedHatProductDataSourceName |} as L
+      WHERE L."timestamp" >= timestamp '{| default .Report.ReportingStart .Report.Inputs.ReportingStart | prestoTimestamp |}'
+          AND L."timestamp" <= timestamp '{| default .Report.ReportingEnd .Report.Inputs.ReportingEnd | prestoTimestamp |}'
+          AND L.dt >= '{| default .Report.ReportingStart .Report.Inputs.ReportingStart | prometheusMetricPartitionFormat |}'
+          AND L.dt <= '{| default .Report.ReportingEnd .Report.Inputs.ReportingEnd | prometheusMetricPartitionFormat |}'
+      GROUP BY L.labels['namespace'], L.labels['pod'], L.labels['label_com_redhat_product'], date_trunc('hour', L."timestamp")
+    ),
+    cte_hourly_pod_by_image AS (
+      SELECT
+          P.labels['pod'] as pod,
+          P.labels['namespace'] as namespace,
+          P.labels['image'] as image,
+          date_trunc('hour', P."timestamp") as interval_start
+      FROM {| dataSourceTableName .Report.Inputs.KubePodContainerInfoRedHatImageDataSourceName |} as P
+      WHERE P."timestamp" >= timestamp '{| default .Report.ReportingStart .Report.Inputs.ReportingStart | prestoTimestamp |}'
+          AND P."timestamp" <= timestamp '{| default .Report.ReportingEnd .Report.Inputs.ReportingEnd | prestoTimestamp |}'
+          AND P.dt >= '{| default .Report.ReportingStart .Report.Inputs.ReportingStart | prometheusMetricPartitionFormat |}'
+          AND P.dt <= '{| default .Report.ReportingEnd .Report.Inputs.ReportingEnd | prometheusMetricPartitionFormat |}'
+      GROUP BY P.labels['namespace'], P.labels['pod'], P.labels['image'], date_trunc('hour', P."timestamp")
+    ),
+    cte_hourly_pod_max_cpu_core_by_label AS (
+      SELECT
+        C.pod,
+        C.namespace,
+        L.product,
+        C.used_cores,
+        C.used_cores * D.cost_per_cpu_core_hour as cost_per_hour,
+        D.currency,
+        C.interval_start
+      FROM cte_hourly_pod_max_cpu_core as C
+      JOIN cte_hourly_pod_by_product_label as L
+        ON C.pod = L.pod
+          AND C.interval_start = L.interval_start
+      JOIN {| dataSourceTableName .Report.Inputs.ProductCpuCorePricingDataSourceName |} as D
+          ON L.product = D.product
+    ),
+    cte_hourly_pod_max_cpu_core_by_image AS (
+      SELECT
+        C.pod,
+        C.namespace,
+        P.product,
+        C.used_cores,
+        C.used_cores * D.cost_per_cpu_core_hour as cost_per_hour,
+        D.currency,
+        C.interval_start
+      FROM cte_hourly_pod_max_cpu_core as C
+      JOIN cte_hourly_pod_by_image as I
+        ON C.pod = I.pod
+          AND C.interval_start = I.interval_start
```

Pretty sure pod names are distinct enough to not be duplicated across namespaces, but since we have namespace we might as well be extra sure

```suggestion
          AND C.namespace = l.namespace
          AND C.interval_start = I.interval_start
```